### PR TITLE
refactor: alert-service

### PR DESCRIPTION
**Pull request by lst97**
# Description

Refactor alert service using SnackBar from Angular

It takes 4 params with 2 optional
type, message, duration, action

Fixes # (issue)

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
## Testing Checklist:

- [x] Tested in latest Chrome
- [x] Tested in latest Safari
- [x] Tested in latest Firefox

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have requested a review from @macite and @jakerenzella on the Pull Request

### CODE REVIEW

**Comment by lst97** [comment]
Screenshot 2022-12-10 at 1 27 37 pm

**Comment by lst97** [comment]
info:
Screenshot 2022-12-10 at 3 27 18 pm
warning:
Screenshot 2022-12-10 at 3 23 00 pm
success:
Screenshot 2022-12-10 at 3 23 22 pm
danger:
Screenshot 2022-12-10 at 3 23 39 pm

**Comment by lst97** [comment]
@jakerenzella @macite

File: src/app/common/services/alert.service.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        this.snackBar.open(message, action, {
+          duration: duration ? duration : 8000,
+          horizontalPosition: 'end',
+          verticalPosition: 'top',
+          panelClass: ['alert-danger'],
+        });
+        break;
+    }
+  }
+}
```
Comment (jakerenzella) [inline src/app/common/services/alert.service.ts]: Can we refactor this to have 4 methods called danger, warning, success and info. This means we no longer need the AlertType type to be known or used in the various callers of the AlertService.

We also only need one private function which opens the snackbar

File: src/styles/common/alert.scss [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  color: white;
+}
+.alert-warning button{
+  background-color: lightsalmon;
+  color: white;
+}
+.alert-danger button{
+  background-color: lightcoral;
+  color: white;
+}
```
Comment (jakerenzella) [inline src/styles/common/alert.scss]: This is good, I'll take a look later and change the colours to better match our style

File: src/app/common/edit-profile-form/edit-profile-form.component.ts [hunk 3]
```diff
@@ -62,16 +62,11 @@ export class EditProfileFormComponent implements OnInit {
...
-          this._snackBar.open('Profile saved', 'dismiss', {
-            duration: 1500,
-            horizontalPosition: 'end',
-            verticalPosition: 'top',
-          });
         }
       },
-      error: (error) => console.log(error),
+      error: (error) => {
+        this.alertService.add(AlertType.DANGER, error);
```
Comment (jakerenzella) [inline src/app/common/edit-profile-form/edit-profile-form.component.ts]: So this would become: `this.alertservice.danger(error)`

**Review by jakerenzella — CHANGES_REQUESTED** [review]
Good start @lst97 - I've left some feedback but this is great. This PR will also need to remove the old alertService code and make sure every component in the repo uses the new service!

File: src/styles/common/alert.scss (hunk 2 above)
Comment (lst97) [reply src/styles/common/alert.scss:16]: commit: https://github.com/doubtfire-lms/doubtfire-web/pull/650/commits/62ac3e891f326d92b52d46bbc9b6208a2e96dbec

File: src/app/common/services/alert.service.ts (hunk 1 above)
Comment (lst97) [reply src/app/common/services/alert.service.ts]: commit: https://github.com/doubtfire-lms/doubtfire-web/pull/650/commits/62ac3e891f326d92b52d46bbc9b6208a2e96dbec

File: src/app/common/edit-profile-form/edit-profile-form.component.ts (hunk 3 above)
Comment (lst97) [reply src/app/common/edit-profile-form/edit-profile-form.component.ts]: commit: https://github.com/doubtfire-lms/doubtfire-web/pull/650/commits/62ac3e891f326d92b52d46bbc9b6208a2e96dbec

**Comment by lst97** [comment]
Working on remove the old alertService code

**Comment by lst97** [comment]
Screenshot 2022-12-11 at 2 14 01 pm

**Comment by lst97** [comment]
Screenshot 2022-12-12 at 2 04 26 pm

**Comment by lst97** [comment]
@jakerenzella our team will refactor all the AngularJS to use new alertSerivce this week.

**Comment by jakerenzella** [comment]
@lst97 before we do that, just a thought, what does a longer, multi-line error message look like with the new component?

**Comment by lst97** [comment]
@jakerenzella 
Screenshot 2022-12-12 at 2 26 31 pm
It will need a new css class

```css
.support-new-line {
  white-space: pre-wrap;
}
```

**Comment by jakerenzella** [comment]
Great! Thanks, go ahead then

**Comment by lst97** [comment]
New alertService can use DURATION.TYPE to set the default timeout, it can also provided a custom timeout. The snackbar will not dismiss if no timeout is provided.

`export const DURATION = {
  INFO: 3000,
  SUCCESS: 5000,
  WARNING: 6000,
  DANGER: 8000,
};`

AngularJS using new alertService must provided a timeout instead.

**Comment by jakerenzella** [comment]
We don't want to pass that in to each component, better to just have a default timeout for each time in the service and an optional ms parameter for the call if the timeout wants to be overridden - think about keeping the service as clean and simple to use as possible

**Comment by lst97** [comment]
> We don't want to pass that in to each component, better to just have a default timeout for each time in the service and an optional ms parameter for the call if the timeout wants to be overridden - think about keeping the service as clean and simple to use as possible

image

I saw this comment in the old component.

I can revert the change if you want @jakerenzella

**Comment by jakerenzella** [comment]
Yes, I don't think that idea is very good - perhaps a negative number can be an unlimited timeout. No timeout should be a default value.

Just get the basic infrastructure in and we can edit from there :) 

Also please remove the default timeout from constants that shouldn't belong there either, if it needs to live somewhere it can stay in the alertService.

Thanks!

**Comment by lst97** [comment]
There are 6 components left to be refactor, one of our team member will handle that. Thank you.

**Comment by ghost** [comment]
@jakerenzella I see there are a alerts.add method which is from the old component alert-list. However the snackbar will dismissed the old alert with the new one, I think we may need to create our own component for new alert list instead, whats your suggestion?

for example, the reason it using alert-list because when the task is over due, it generate two alerts, one for the indication of successful upload, the other for "You have submitted after the deadline for feedback. Your task will not be reviewed by a tutor. It is now your responsibility to ensure this task meets the required standard."

image

**Comment by lst97** [comment]
Or we can add a timeout.
```typescript
 message.forEach( (message, index) => {
    setTimeout(() => {
      
        this.snackBar.open(message.text, action, {
            duration: this.timeOut,
            verticalPosition: 'top',
            horizontalPosition: 'end',
            panelClass: [allow-newline],
        });
        

    }, index * (this.timeOut+500)); // 500 => timeout between two messages
```

**Pull request closed without merging**